### ISSUE TITLE
Refactor Wazuh framework search, sort, offset and limit post processing

### ISSUE THREAD

**Issue by davidjiglesias** [wazuh/wazuh#3684]
Hi team, 

I have found the need to refactor the actual data post processing used at framework level. There is a lot of code duplication that can be eliminated with this approach.

Right now, framework functions are applying search, sort, offset and limit processing to the output, such as:
```python
        if search:
            req = search_array(req, search['value'], search['negation'])

        if sort:
            req = sort_array(req, order=sort['order'])
        else:
            req = sort_array(req)

        return {'items': cut_array(req, offset, limit), 'totalItems': len(req)}
```
The use of a common `utils.py` function to take care of all the processing is proposed, and the solution would look like:
```python
        return process_array(groups, search_text=search_text, search_in_fields=search_in_fields,
                             complementary_search=complementary_search, sort_by=sort_by, sort_ascending=sort_ascending,
                             offset=offset, limit=limit)
```
All the `process_array` parameters are optional except for data|array.
```python
def process_array(array, search_text=None, complementary_search=False, search_in_fields=None, sort_by=None,
                  sort_ascending=True, allowed_sort_fields=None, offset=0, limit=None):
    """ Process a Wazuh framework data array

    :param array: Array to process
    :param search_text: Text to search and search type
    :param complementary_search: Perform a complementary search
    :param search_in_fields: Fields to search in
    :param sort_by: Fields to sort_by. Will sort the array directly if [''] is received
    :param sort_ascending: Sort order ascending or descending
    :param allowed_sort_fields: Allowed fields to sort_by
    :param offset: First element to return.
    :param limit: Maximum number of elements to return
    :return: Dictionary: {'items': Processed array, 'totalItems': Number of items, before applying offset and limit)}
    """
```

**Comment by davidjiglesias** [comment]
Status update
## Summary
Today I have completed the refactoring of `search_array`, `sort_array` and `cut_array` post processing in framework modules. I have also updated integration tests when neccesary and fixed an error in lists endpoints.

## Tasks
* Fix `GET /lists/files` controller to correctly apply `search`, `sort`, `offset` and `limit` parameters
* Update spec for previous fix in `GET /lists/files` endpoint
* Update `lists` integration tests
* Update `rules` integration tests
* Refactor `agent.py`
* Refactor `cdb_list.py`
* Refactor `server.py`
* Refactor `decoder.py`
* Refactor `manager.py`
* Refactor `rule.py`
* Add `process_array` to`utils.py`

## Tests
Agent tests (Changes affect GET /agents/groups and GET/agents/groups/{group_id}/files)
```
http://localhost:55000/agents/groups?pretty=false&wait_for_complete=false&sort=-count
{
  "data": {
    "items": [
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 12,
        "mergedSum": "c6309ff81a74781f6b55b68129a76738",
        "name": "default"
      },
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 6,
        "mergedSum": "13a2d4d772bccb56646b4ca9d1dc6689",
        "name": "group2"
      },
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 6,
        "mergedSum": "f5e0305e248bd00e05445f329bd1dd5e",
        "name": "group1"
      },
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 5,
        "mergedSum": "49f64bf1678a3b57b07bbc09fc62e6ba",
        "name": "group3"
      }
    ],
    "totalItems": 4
  }
}

http://localhost:55000/agents/groups?pretty=false&wait_for_complete=false&limit=1
{
  "data": {
    "items": [
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 12,
        "mergedSum": "c6309ff81a74781f6b55b68129a76738",
        "name": "default"
      }
    ],
    "totalItems": 4
  }
}

http://localhost:55000/agents/groups/group1/files?pretty=false&wait_for_complete=false&offset=0&limit=500&search=ar
{
  "data": {
    "items": [
      {
        "filename": "ar.conf",
        "hash": "76d8be9b97d8eae4c239e530ee7e71c8"
      }
    ],
    "totalItems": 1
  }
}

http://localhost:55000/agents/groups/group1/files?pretty=false&wait_for_complete=false&offset=0&limit=500&sort=-filename
{
  "data": {
    "items": [
      {
        "filename": "merged.mg",
        "hash": "f5e0305e248bd00e05445f329bd1dd5e"
      },
      {
        "filename": "ar.conf",
        "hash": "76d8be9b97d8eae4c239e530ee7e71c8"
      },
      {
        "filename": "agent.conf",
        "hash": "ab73af41699f13fdd81903b5f23d8d00"
      }
    ],
    "totalItems": 3
  }
}
```
Lists tests:
```
test/test_lists_endpoints.tavern.yaml::GET /lists
  /var/ossec/framework/python/lib/python3.7/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================== 3 passed, 2 warnings in 0.92 seconds ===================================
```
Server tests (Used a broadcasted request)
```
http://localhost:55000/syscheck?pretty=false&wait_for_complete=false
{
  "message": "Restarting Syscheck/Rootcheck on all agents"
}
```
Decoder tests:
```
=========================================== test session starts ===========================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /var/ossec/framework/python/bin
plugins: cov-2.7.1, tavern-0.26.4
collected 23 items                                                                                        

test/test_decoders_endpoints.tavern.yaml::GET /decoders PASSED                                      [  4%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[file] PASSED                                [  8%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[path] PASSED                                [ 13%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[name] PASSED                                [ 17%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[position] PASSED                            [ 21%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[status] PASSED                              [ 26%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name} PASSED                       [ 30%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[file] PASSED                 [ 34%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[path] PASSED                 [ 39%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[name] PASSED                 [ 43%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[position] PASSED             [ 47%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[status] PASSED               [ 52%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files PASSED                                [ 56%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files[file] PASSED                          [ 60%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files[path] PASSED                          [ 65%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files[status] PASSED                        [ 69%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents PASSED                              [ 73%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[file] PASSED                        [ 78%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[name] PASSED                        [ 82%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[path] PASSED                        [ 86%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[position] PASSED                    [ 91%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[status] PASSED                      [ 95%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files/{file}/download PASSED                [100%]

============================================ warnings summary =============================================
/var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126
  /var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    if not isinstance(key, collections.Hashable):

test/test_decoders_endpoints.tavern.yaml::GET /decoders
  /var/ossec/framework/python/lib/python3.7/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================== 23 passed, 2 warnings in 3.81 seconds ==================================
```
Manager tests: (Changes affect `GET /manager/logs`)
```
http://localhost:55000/manager/logs?pretty=false&wait_for_complete=false&offset=0&limit=500&search=6008
{
  "data": {
    "items": [
      {
        "description": " (6008): File integrity monitoring scan started.",
        "level": "info",
        "tag": "ossec-syscheckd",
        "timestamp": "2019-07-19T10:55:51Z"
      },
      {
        "description": " (6008): File integrity monitoring scan started.",
        "level": "info",
        "tag": "ossec-syscheckd",
        "timestamp": "2019-07-19T10:54:55Z"
      }
    ],
    "totalItems": 2
  }
}

http://localhost:55000/manager/logs?pretty=false&wait_for_complete=false&offset=0&limit=3&sort=timestamp
{
  "data": {
    "items": [
      {
        "description": " Remote syslog server not configured. Clean exit.",
        "level": "info",
        "tag": "ossec-csyslogd",
        "timestamp": "2019-07-17T09:56:40Z"
      },
      {
        "description": " Database not configured. Clean exit.",
        "level": "info",
        "tag": "ossec-dbd",
        "timestamp": "2019-07-17T09:56:40Z"
      },
      {
        "description": " Remote integrations not configured. Clean exit.",
        "level": "info",
        "tag": "ossec-integratord",
        "timestamp": "2019-07-17T09:56:40Z"
      }
    ],
    "totalItems": 171
  }
}
```
Rule tests:
```
========================================== test session starts ===========================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /var/ossec/framework/python/bin
plugins: cov-2.7.1, tavern-0.26.4
collected 10 items                                                                                        

test/test_rules_endpoint.tavern.yaml::GET /rules PASSED                                             [ 10%]
test/test_rules_endpoint.tavern.yaml::GET /rules filters PASSED                                     [ 20%]
test/test_rules_endpoint.tavern.yaml::GET /rules/pci PASSED                                         [ 30%]
test/test_rules_endpoint.tavern.yaml::GET /rules/gdpr PASSED                                        [ 40%]
test/test_rules_endpoint.tavern.yaml::GET /rules/gpg13 PASSED                                       [ 50%]
test/test_rules_endpoint.tavern.yaml::GET /rules/hipaa PASSED                                       [ 60%]
test/test_rules_endpoint.tavern.yaml::GET /rules/nist-800-53 PASSED                                 [ 70%]
test/test_rules_endpoint.tavern.yaml::GET /rules/groups PASSED                                      [ 80%]
test/test_rules_endpoint.tavern.yaml::GET /rules/files PASSED                                       [ 90%]
test/test_rules_endpoint.tavern.yaml::GET /rules/{id} PASSED                                        [100%]

============================================ warnings summary =============================================
/var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126
  /var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    if not isinstance(key, collections.Hashable):

test/test_rules_endpoint.tavern.yaml::GET /rules
test/test_rules_endpoint.tavern.yaml::GET /rules/pci
test/test_rules_endpoint.tavern.yaml::GET /rules/gdpr
test/test_rules_endpoint.tavern.yaml::GET /rules/gpg13
test/test_rules_endpoint.tavern.yaml::GET /rules/hipaa
test/test_rules_endpoint.tavern.yaml::GET /rules/nist-800-53
test/test_rules_endpoint.tavern.yaml::GET /rules/groups
test/test_rules_endpoint.tavern.yaml::GET /rules/files
test/test_rules_endpoint.tavern.yaml::GET /rules/{id}
  /var/ossec/framework/python/lib/python3.7/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================= 10 passed, 10 warnings in 15.75 seconds =================================
```
## Pending

**Comment by davidjiglesias** [comment]
Hi team, 

After further review, we have arrived to the conclusion that actual framework data post-processing(search,sort,offset and limit) is extremely dependent and coupled with API needs. We are working on a solution to continue refactoring and to also decouple framework and API in this regard.

**Comment by davidjiglesias** [comment]
Status update
## Summary
Today I have completed the refactoring of `search_array`, `sort_array` and `cut_array` post processing in framework modules. I have also updated integration tests when neccesary and fixed an error in lists endpoints.

## Tasks
* Fix `GET /lists/files` controller to correctly apply `search`, `sort`, `offset` and `limit` parameters bc50afdda4f01c58c7f87b05175598985dd68206
* Update spec for previous fix in `GET /lists/files` endpoint 5cdec3fa270bdcb157114223d48c866a68e029f7
* Update `lists` integration tests 4a012c40e032fd045015182948c74947bf43ee13
* Update `rules` integration tests cfbd7ec1928d9b0cbf9a6c57784f56438df1479a
* Refactor `agent.py` 056e6a0f248e260ab3c1e3f68db0703edc8582ae
* Refactor `cdb_list.py` c330de0cd2fd101106df0f3f03956b8cecf8e127
* Refactor `server.py` 5fa9005ece9f85f88eb06596f50cb9712865e286
* Refactor `decoder.py` aa45cf410cb095c3615be1c1f31a138e3629a1f5
* Refactor `manager.py` 8d90446e91a2b1fa6439cc0c824688c788cd0ce7
* Refactor `rule.py` f1d420d5b28c69a6d8ed02b3ce4303d8ab7bfe82
* Add `process_array` to`utils.py` be6232e9d71997e20787004271c35eccbdd74f05

## Tests
Agent tests (Changes affect `GET /agents/groups` and `GET/agents/groups/{group_id}/files`)
```
http://localhost:55000/agents/groups?pretty=false&wait_for_complete=false&sort=-count
{
  "data": {
    "items": [
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 12,
        "mergedSum": "c6309ff81a74781f6b55b68129a76738",
        "name": "default"
      },
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 6,
        "mergedSum": "13a2d4d772bccb56646b4ca9d1dc6689",
        "name": "group2"
      },
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 6,
        "mergedSum": "f5e0305e248bd00e05445f329bd1dd5e",
        "name": "group1"
      },
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 5,
        "mergedSum": "49f64bf1678a3b57b07bbc09fc62e6ba",
        "name": "group3"
      }
    ],
    "totalItems": 4
  }
}

http://localhost:55000/agents/groups?pretty=false&wait_for_complete=false&limit=1
{
  "data": {
    "items": [
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "count": 12,
        "mergedSum": "c6309ff81a74781f6b55b68129a76738",
        "name": "default"
      }
    ],
    "totalItems": 4
  }
}

http://localhost:55000/agents/groups/group1/files?pretty=false&wait_for_complete=false&offset=0&limit=500&search=ar
{
  "data": {
    "items": [
      {
        "filename": "ar.conf",
        "hash": "76d8be9b97d8eae4c239e530ee7e71c8"
      }
    ],
    "totalItems": 1
  }
}

http://localhost:55000/agents/groups/group1/files?pretty=false&wait_for_complete=false&offset=0&limit=500&sort=-filename
{
  "data": {
    "items": [
      {
        "filename": "merged.mg",
        "hash": "f5e0305e248bd00e05445f329bd1dd5e"
      },
      {
        "filename": "ar.conf",
        "hash": "76d8be9b97d8eae4c239e530ee7e71c8"
      },
      {
        "filename": "agent.conf",
        "hash": "ab73af41699f13fdd81903b5f23d8d00"
      }
    ],
    "totalItems": 3
  }
}
```
Lists tests:
```
test/test_lists_endpoints.tavern.yaml::GET /lists
  /var/ossec/framework/python/lib/python3.7/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================== 3 passed, 2 warnings in 0.92 seconds ===================================
```
Server tests (Used a broadcasted request)
```
http://localhost:55000/syscheck?pretty=false&wait_for_complete=false
{
  "message": "Restarting Syscheck/Rootcheck on all agents"
}
```
Decoder tests:
```
=========================================== test session starts ===========================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /var/ossec/framework/python/bin
plugins: cov-2.7.1, tavern-0.26.4
collected 23 items                                                                                        

test/test_decoders_endpoints.tavern.yaml::GET /decoders PASSED                                      [  4%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[file] PASSED                                [  8%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[path] PASSED                                [ 13%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[name] PASSED                                [ 17%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[position] PASSED                            [ 21%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders[status] PASSED                              [ 26%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name} PASSED                       [ 30%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[file] PASSED                 [ 34%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[path] PASSED                 [ 39%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[name] PASSED                 [ 43%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[position] PASSED             [ 47%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/{decoder_name}[status] PASSED               [ 52%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files PASSED                                [ 56%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files[file] PASSED                          [ 60%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files[path] PASSED                          [ 65%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files[status] PASSED                        [ 69%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents PASSED                              [ 73%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[file] PASSED                        [ 78%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[name] PASSED                        [ 82%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[path] PASSED                        [ 86%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[position] PASSED                    [ 91%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/parents[status] PASSED                      [ 95%]
test/test_decoders_endpoints.tavern.yaml::GET /decoders/files/{file}/download PASSED                [100%]

============================================ warnings summary =============================================
/var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126
  /var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    if not isinstance(key, collections.Hashable):

test/test_decoders_endpoints.tavern.yaml::GET /decoders
  /var/ossec/framework/python/lib/python3.7/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================== 23 passed, 2 warnings in 3.81 seconds ==================================
```
Manager tests: (Changes affect `GET /manager/logs`)
```
http://localhost:55000/manager/logs?pretty=false&wait_for_complete=false&offset=0&limit=500&search=6008
{
  "data": {
    "items": [
      {
        "description": " (6008): File integrity monitoring scan started.",
        "level": "info",
        "tag": "ossec-syscheckd",
        "timestamp": "2019-07-19T10:55:51Z"
      },
      {
        "description": " (6008): File integrity monitoring scan started.",
        "level": "info",
        "tag": "ossec-syscheckd",
        "timestamp": "2019-07-19T10:54:55Z"
      }
    ],
    "totalItems": 2
  }
}

http://localhost:55000/manager/logs?pretty=false&wait_for_complete=false&offset=0&limit=3&sort=timestamp
{
  "data": {
    "items": [
      {
        "description": " Remote syslog server not configured. Clean exit.",
        "level": "info",
        "tag": "ossec-csyslogd",
        "timestamp": "2019-07-17T09:56:40Z"
      },
      {
        "description": " Database not configured. Clean exit.",
        "level": "info",
        "tag": "ossec-dbd",
        "timestamp": "2019-07-17T09:56:40Z"
      },
      {
        "description": " Remote integrations not configured. Clean exit.",
        "level": "info",
        "tag": "ossec-integratord",
        "timestamp": "2019-07-17T09:56:40Z"
      }
    ],
    "totalItems": 171
  }
}
```
Rule tests:
```
========================================== test session starts ===========================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /var/ossec/framework/python/bin
plugins: cov-2.7.1, tavern-0.26.4
collected 10 items                                                                                        

test/test_rules_endpoint.tavern.yaml::GET /rules PASSED                                             [ 10%]
test/test_rules_endpoint.tavern.yaml::GET /rules filters PASSED                                     [ 20%]
test/test_rules_endpoint.tavern.yaml::GET /rules/pci PASSED                                         [ 30%]
test/test_rules_endpoint.tavern.yaml::GET /rules/gdpr PASSED                                        [ 40%]
test/test_rules_endpoint.tavern.yaml::GET /rules/gpg13 PASSED                                       [ 50%]
test/test_rules_endpoint.tavern.yaml::GET /rules/hipaa PASSED                                       [ 60%]
test/test_rules_endpoint.tavern.yaml::GET /rules/nist-800-53 PASSED                                 [ 70%]
test/test_rules_endpoint.tavern.yaml::GET /rules/groups PASSED                                      [ 80%]
test/test_rules_endpoint.tavern.yaml::GET /rules/files PASSED                                       [ 90%]
test/test_rules_endpoint.tavern.yaml::GET /rules/{id} PASSED                                        [100%]

============================================ warnings summary =============================================
/var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126
  /var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    if not isinstance(key, collections.Hashable):

test/test_rules_endpoint.tavern.yaml::GET /rules
test/test_rules_endpoint.tavern.yaml::GET /rules/pci
test/test_rules_endpoint.tavern.yaml::GET /rules/gdpr
test/test_rules_endpoint.tavern.yaml::GET /rules/gpg13
test/test_rules_endpoint.tavern.yaml::GET /rules/hipaa
test/test_rules_endpoint.tavern.yaml::GET /rules/nist-800-53
test/test_rules_endpoint.tavern.yaml::GET /rules/groups
test/test_rules_endpoint.tavern.yaml::GET /rules/files
test/test_rules_endpoint.tavern.yaml::GET /rules/{id}
  /var/ossec/framework/python/lib/python3.7/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================= 10 passed, 10 warnings in 15.75 seconds =================================
```
## Pending

**Comment by davidjiglesias** [comment]
Status update
## Summary
Today I have completed the refactoring of controllers and framework modules to decouple data post-processing.

## Tasks
* Update controllers to send appropiate decoupled information to framework about formatting 2accca3447542ba9e53c57ac075f206af9ef81d6
* Update framework to always expect the same input from any source eb683bfc9270403b240b0d30f6de271604762e8e
* Update exception code 31233fda3f6766785d41206451731d3b4004bc92
* Update process_array refactor function 5b538416848fdfe3124ce1eddc0b08fc974ebd1a

## Tests

## Pending
